File: quant-trading-system/src/api/v1/endpoints/monitoring.py
"""
实时监控API

提供交易监控相关的WebSocket和HTTP接口
"""
import asyncio
from datetime import datetime
from typing import Dict, List, Optional, Set
from fastapi import APIRouter, WebSocket, WebSocketDisconnect, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession

from src.models.base import get_db
from src.models.order import Order
from src.models.enums import OrderStatus
from src.services.order_service import OrderService
from src.services.account_service import AccountService
from src.strategy.manager import strategy_manager
from src.common.logger import TradingLogger
from src.market_data.manager import MarketDataManager

logger = TradingLogger(__name__)
router = APIRouter()


class MonitoringManager:
    """
    监控管理器

    管理WebSocket连接和实时数据分发
    """

    def __init__(self):
        self.active_connections: Set[WebSocket] = set()
        self._running = False
        self._broadcast_task: Optional[asyncio.Task] = None

    async def connect(self, websocket: WebSocket):
        """接受新的WebSocket连接"""
        await websocket.accept()
        self.active_connections.add(websocket)
        logger.info(f"监控连接已建立，当前连接数: {len(self.active_connections)}")

    def disconnect(self, websocket: WebSocket):
        """断开WebSocket连接"""
        self.active_connections.discard(websocket)
        logger.info(f"监控连接已断开，当前连接数: {len(self.active_connections)}")

    async def broadcast(self, message: Dict):
        """广播消息到所有连接"""
        if not self.active_connections:
            return

        disconnected = set()
        for connection in self.active_connections:
            try:
                await connection.send_json(message)
            except Exception:
                disconnected.add(connection)

        # 清理断开的连接
        for conn in disconnected:
            self.active_connections.discard(conn)

    async def send_personal_message(self, message: Dict, websocket: WebSocket):
        """发送个人消息"""
        try:
            await websocket.send_json(message)
        except Exception as e:
            logger.error(f"发送消息失败: {e}")


# 全局监控管理器实例
monitoring_manager = MonitoringManager()


@router.websocket("/ws")
async def monitoring_websocket(
    websocket: WebSocket,
    db: AsyncSession = Depends(get_db)
):
    """
    实时监控WebSocket

    提供实时数据流：
    - 订单状态更新
    - 持仓变化
    - 策略信号
    - 风控告警
    """
    await monitoring_manager.connect(websocket)

    try:
        # 发送初始连接成功消息
        await websocket.send_json({
            "type": "connected",
            "message": "监控连接已建立",
            "timestamp": datetime.now().isoformat()
        })

        while True:
            # 接收客户端消息
            data = await websocket.receive_json()
            action = data.get("action")

            if action == "subscribe":
                # 处理订阅请求
                symbols = data.get("symbols", [])
                await websocket.send_json({
                    "type": "subscribed",
                    "symbols": symbols
                })

            elif action == "ping":
                await websocket.send_json({"type": "pong"})

            elif action == "get_portfolio":
                # 获取组合信息
                account_id = data.get("account_id", 1)
                portfolio = await get_portfolio_summary(db, account_id)
                await websocket.send_json({
                    "type": "portfolio_update",
                    "data": portfolio
                })

            elif action == "get_positions":
                # 获取持仓列表
                account_id = data.get("account_id", 1)
                positions = await get_positions_data(db, account_id)
                await websocket.send_json({
                    "type": "positions_update",
                    "data": positions
                })

            elif action == "get_active_orders":
                # 获取活跃订单
                account_id = data.get("account_id", 1)
                orders = await get_active_orders_data(db, account_id)
                await websocket.send_json({
                    "type": "orders_update",
                    "data": orders
                })

            elif action == "get_strategies":
                # 获取策略状态
                strategies = get_strategies_status()
                await websocket.send_json({
                    "type": "strategies_update",
                    "data": strategies
                })

    except WebSocketDisconnect:
        monitoring_manager.disconnect(websocket)
    except Exception as e:
        logger.error(f"WebSocket错误: {e}")
        monitoring_manager.disconnect(websocket)


async def get_portfolio_summary(db: AsyncSession, account_id: int) -> Dict:
    """获取组合摘要"""
    try:
        account_service = AccountService(db)
        account = await account_service.get_account(account_id)

        if not account:
            return {
                "total_value": 0,
                "cash": 0,
                "position_value": 0,
                "daily_pnl": 0,
                "daily_return": 0
            }

        return {
            "total_value": float(account.total_equity),
            "cash": float(account.available),
            "position_value": float(account.market_value),
            "daily_pnl": float(account.unrealized_pnl),
            "daily_return": float(account.total_return_pct) * 100
        }
    except Exception as e:
        logger.error(f"获取组合摘要失败: {e}")
        return {
            "total_value": 0,
            "cash": 0,
            "position_value": 0,
            "daily_pnl": 0,
            "daily_return": 0
        }


async def get_positions_data(db: AsyncSession, account_id: int) -> List[Dict]:
    """获取持仓数据"""
    try:
        account_service = AccountService(db)
        positions = await account_service.get_positions(account_id)

        result = []
        for pos in positions:
            # 计算盈亏
            pnl_pct = 0
            if pos.avg_price and pos.avg_price > 0:
                pnl_pct = ((pos.current_price - pos.avg_price) / pos.avg_price) * 100

            result.append({
                "symbol": pos.symbol,
                "name": pos.symbol_name or pos.symbol,
                "qty": pos.qty,
                "cost": float(pos.avg_price) if pos.avg_price else 0,
                "price": float(pos.current_price) if pos.current_price else 0,
                "pnl": round(pnl_pct, 2),
                "market_value": float(pos.market_value) if pos.market_value else 0,
                "alerts": []  # 可由风控系统填充
            })

        return result
    except Exception as e:
        logger.error(f"获取持仓数据失败: {e}")
        return []


async def get_active_orders_data(db: AsyncSession, account_id: int) -> List[Dict]:
    """获取活跃订单数据"""
    try:
        order_service = OrderService(db)
        orders = await order_service.get_active_orders(account_id)

        result = []
        for order in orders:
            result.append({
                "order_id": order.order_id,
                "symbol": order.symbol,
                "direction": order.direction.value,
                "qty": order.qty,
                "filled_qty": order.filled_qty,
                "price": float(order.price) if order.price else None,
                "status": order.status.value,
                "created_at": order.created_at.isoformat() if order.created_at else None
            })

        return result
    except Exception as e:
        logger.error(f"获取订单数据失败: {e}")
        return []


def get_strategies_status() -> List[Dict]:
    """获取策略状态"""
    try:
        strategies = []
        for strategy_id in strategy_manager.list_strategies():
            strategy = strategy_manager.get_strategy(strategy_id)
            if strategy:
                stats = strategy.stats
                strategies.append({
                    "id": strategy_id,
                    "name": strategy.name,
                    "type": strategy.__class__.__name__,
                    "status": stats.get("state", "unknown"),
                    "performance": {
                        "return": 0,  # 从策略统计中获取
                        "sharpe": 0
                    }
                })
        return strategies
    except Exception as e:
        logger.error(f"获取策略状态失败: {e}")
        return []


@router.get("/portfolio/{account_id}")
async def get_portfolio(
    account_id: int,
    db: AsyncSession = Depends(get_db)
):
    """获取账户组合信息"""
    return await get_portfolio_summary(db, account_id)


@router.get("/positions/{account_id}")
async def get_positions(
    account_id: int,
    db: AsyncSession = Depends(get_db)
):
    """获取持仓列表"""
    return await get_positions_data(db, account_id)


@router.get("/active-orders/{account_id}")
async def get_active_orders(
    account_id: int,
    db: AsyncSession = Depends(get_db)
):
    """获取活跃订单"""
    return await get_active_orders_data(db, account_id)


@router.get("/strategies")
async def get_strategies():
    """获取策略列表"""
    return get_strategies_status()


@router.get("/risk-metrics/{account_id}")
async def get_risk_metrics(account_id: int):
    """获取风控指标"""
    # 这里可以从风控系统获取实际数据
    return [
        {
            "name": "单票仓位",
            "value": "12.5%",
            "limit": "20%",
            "percent": 62.5,
            "status": "normal"
        },
        {
            "name": "总仓位",
            "value": "45.2%",
            "limit": "60%",
            "percent": 75.3,
            "status": "warning"
        },
        {
            "name": "日亏损",
            "value": "0.8%",
            "limit": "2%",
            "percent": 40,
            "status": "normal"
        }
    ]

File: quant-trading-system/web/src/stores/monitoring.ts
/**
 * 实时监控Store
 *
 * 管理交易监控相关的状态和WebSocket连接
 */
import { defineStore } from 'pinia'
import { ref, computed } from 'vue'

// API基础URL
const API_BASE = 'http://localhost:9000/api/v1'
const WS_URL = 'ws://localhost:9000/api/v1/monitoring/ws'

// 类型定义
export interface PortfolioData {
  total_value: number
  cash: number
  position_value: number
  daily_pnl: number
  daily_return: number
}

export interface PositionData {
  symbol: string
  name: string
  qty: number
  cost: number
  price: number
  pnl: number
  market_value: number
  alerts?: AlertItem[]
}

export interface OrderData {
  order_id: string
  symbol: string
  direction: 'BUY' | 'SELL'
  qty: number
  filled_qty: number
  price?: number
  status: string
  created_at: string
}

export interface StrategyData {
  id: string
  name: string
  type: string
  status: 'active' | 'paused' | 'stopped'
  performance: {
    return: number
    sharpe: number
  }
}

export interface RiskMetric {
  name: string
  value: string
  limit: string
  percent: number
  status: 'normal' | 'warning' | 'danger'
}

export interface AlertItem {
  type: 'warning' | 'danger' | 'info'
  message: string
}

export interface SignalData {
  id: string
  type: 'buy' | 'sell' | 'risk' | 'stop'
  symbol: string
  name: string
  strategy: string
  price: number
  time: string
}

export interface LogData {
  id: string
  time: string
  level: 'INFO' | 'WARN' | 'ERROR'
  message: string
}

export const useMonitoringStore = defineStore('monitoring', () => {
  // ============ State ============

  // WebSocket连接状态
  const wsConnected = ref(false)
  const wsConnecting = ref(false)
  const wsError = ref<string | null>(null)

  // 监控数据
  const portfolio = ref<PortfolioData>({
    total_value: 0,
    cash: 0,
    position_value: 0,
    daily_pnl: 0,
    daily_return: 0
  })

  const positions = ref<PositionData[]>([])
  const activeOrders = ref<OrderData[]>([])
  const strategies = ref<StrategyData[]>([])
  const riskMetrics = ref<RiskMetric[]>([])
  const signals = ref<SignalData[]>([])
  const logs = ref<LogData[]>([])

  // 系统状态
  const isTrading = ref(false)
  const selectedAccountId = ref<number>(1)

  // WebSocket实例
  let ws: WebSocket | null = null
  let reconnectTimer: ReturnType<typeof setTimeout> | null = null
  let heartbeatTimer: ReturnType<typeof setInterval> | null = null

  // ============ Getters ============

  const isConnected = computed(() => wsConnected.value)

  const positionCount = computed(() => positions.value.length)

  const activeOrderCount = computed(() => activeOrders.value.length)

  const totalPnl = computed(() => {
    return positions.value.reduce((sum, pos) => {
      return sum + (pos.pnl * pos.market_value / 100)
    }, 0)
  })

  const riskStatus = computed(() => {
    const hasDanger = riskMetrics.value.some(m => m.status === 'danger')
    const hasWarning = riskMetrics.value.some(m => m.status === 'warning')
    if (hasDanger) return { text: '告警', class: 'danger' }
    if (hasWarning) return { text: '注意', class: 'warning' }
    return { text: '正常', class: 'success' }
  })

  // ============ Actions ============

  // 连接WebSocket
  async function connect() {
    if (wsConnecting.value || wsConnected.value) return

    wsConnecting.value = true
    wsError.value = null

    try {
      ws = new WebSocket(WS_URL)

      ws.onopen = () => {
        console.log('[Monitoring] WebSocket connected')
        wsConnected.value = true
        wsConnecting.value = false

        // 启动心跳
        startHeartbeat()

        // 订阅数据
        subscribeData()
      }

      ws.onmessage = (event) => {
        try {
          const message = JSON.parse(event.data)
          handleMessage(message)
        } catch (err) {
          console.error('[Monitoring] Failed to parse message:', err)
        }
      }

      ws.onclose = () => {
        console.log('[Monitoring] WebSocket disconnected')
        wsConnected.value = false
        wsConnecting.value = false
        stopHeartbeat()
        scheduleReconnect()
      }

      ws.onerror = (error) => {
        console.error('[Monitoring] WebSocket error:', error)
        wsError.value = 'WebSocket连接错误'
        wsConnecting.value = false
      }
    } catch (err) {
      console.error('[Monitoring] Failed to connect:', err)
      wsError.value = '连接失败'
      wsConnecting.value = false
      scheduleReconnect()
    }
  }

  // 断开连接
  function disconnect() {
    if (reconnectTimer) {
      clearTimeout(reconnectTimer)
      reconnectTimer = null
    }
    stopHeartbeat()

    if (ws) {
      ws.close()
      ws = null
    }

    wsConnected.value = false
  }

  // 处理收到的消息
  function handleMessage(message: any) {
    switch (message.type) {
      case 'connected':
        console.log('[Monitoring]', message.message)
        break

      case 'portfolio_update':
        portfolio.value = message.data
        break

      case 'positions_update':
        positions.value = message.data
        break

      case 'orders_update':
        activeOrders.value = message.data
        break

      case 'strategies_update':
        strategies.value = message.data
        break

      case 'risk_update':
        riskMetrics.value = message.data
        break

      case 'signal':
        signals.value.unshift(message.data)
        if (signals.value.length > 50) {
          signals.value = signals.value.slice(0, 50)
        }
        break

      case 'log':
        logs.value.unshift(message.data)
        if (logs.value.length > 100) {
          logs.value = logs.value.slice(0, 100)
        }
        break

      case 'pong':
        // 心跳响应
        break

      case 'error':
        console.error('[Monitoring] Server error:', message.message)
        wsError.value = message.message
        break

      default:
        console.log('[Monitoring] Unknown message type:', message.type)
    }
  }

  // 订阅数据
  function subscribeData() {
    if (!ws || ws.readyState !== WebSocket.OPEN) return

    ws.send(JSON.stringify({
      action: 'get_portfolio',
      account_id: selectedAccountId.value
    }))

    ws.send(JSON.stringify({
      action: 'get_positions',
      account_id: selectedAccountId.value
    }))

    ws.send(JSON.stringify({
      action: 'get_active_orders',
      account_id: selectedAccountId.value
    }))

    ws.send(JSON.stringify({
      action: 'get_strategies'
    }))
  }

  // 启动心跳
  function startHeartbeat() {
    heartbeatTimer = setInterval(() => {
      if (ws && ws.readyState === WebSocket.OPEN) {
        ws.send(JSON.stringify({ action: 'ping' }))
      }
    }, 30000)
  }

  // 停止心跳
  function stopHeartbeat() {
    if (heartbeatTimer) {
      clearInterval(heartbeatTimer)
      heartbeatTimer = null
    }
  }

  // 计划重连
  function scheduleReconnect() {
    if (reconnectTimer) return

    console.log('[Monitoring] Reconnecting in 5s...')
    reconnectTimer = setTimeout(() => {
      reconnectTimer = null
      connect()
    }, 5000)
  }

  // 刷新数据
  async function refreshData() {
    subscribeData()
  }

  // 切换交易状态
  function toggleTrading() {
    isTrading.value = !isTrading.value
    // 可以发送命令到后端
  }

  // 设置选中账户
  function setAccountId(accountId: number) {
    selectedAccountId.value = accountId
    refreshData()
  }

  // 添加模拟数据（用于测试）
  function addMockData() {
    portfolio.value = {
      total_value: 1245890.50,
      cash: 456890.30,
      position_value: 789000.20,
      daily_pnl: 29245.80,
      daily_return: 2.35
    }

    positions.value = [
      { symbol: '600519.SH', name: '贵州茅台', qty: 100, cost: 1650.00, price: 1685.00, pnl: 2.12, market_value: 168500 },
      { symbol: '000858.SZ', name: '五粮液', qty: 200, cost: 145.00, price: 142.30, pnl: -1.86, market_value: 28460 },
      { symbol: '000333.SZ', name: '美的集团', qty: 500, cost: 64.50, price: 65.80, pnl: 2.02, market_value: 32900 }
    ]

    activeOrders.value = [
      { order_id: 'ORD001', symbol: '600519.SH', direction: 'BUY', qty: 100, filled_qty: 0, price: 1680.00, status: 'PENDING', created_at: new Date().toISOString() }
    ]

    strategies.value = [
      { id: '1', name: '双均线突破', type: 'MACross', status: 'active', performance: { return: 15.2, sharpe: 1.5 } },
      { id: '2', name: 'MACD金叉', type: 'MACD', status: 'active', performance: { return: 8.5, sharpe: 1.2 } }
    ]

    riskMetrics.value = [
      { name: '单票仓位', value: '12.5%', limit: '20%', percent: 62.5, status: 'normal' },
      { name: '总仓位', value: '45.2%', limit: '60%', percent: 75.3, status: 'warning' },
      { name: '日亏损', value: '0.8%', limit: '2%', percent: 40, status: 'normal' }
    ]

    signals.value = [
      { id: '1', type: 'buy', symbol: '600519.SH', name: '贵州茅台', strategy: '双均线突破', price: 1685.00, time: '09:45:23' }
    ]

    logs.value = [
      { id: '1', time: '09:45:23', level: 'INFO', message: '买入信号触发: 600519.SH @ 1685.00' },
      { id: '2', time: '09:42:15', level: 'WARN', message: '卖出信号触发: 300750.SZ @ 215.50' }
    ]
  }

  return {
    // State
    wsConnected,
    wsConnecting,
    wsError,
    portfolio,
    positions,
    activeOrders,
    strategies,
    riskMetrics,
    signals,
    logs,
    isTrading,
    selectedAccountId,

    // Getters
    isConnected,
    positionCount,
    activeOrderCount,
    totalPnl,
    riskStatus,

    // Actions
    connect,
    disconnect,
    refreshData,
    toggleTrading,
    setAccountId,
    addMockData
  }
})
